### ISSUE TITLE
Proposal for minimizing/removing secrets in the Terraform state file

### ISSUE THREAD

**Issue by rwblokzijl** [hashicorp/terraform#34860]
### Terraform Version

```shell
v1.7.5
```

### Use Cases

With the additional benefit of low privilege Terraform plans.

# Background:

Terraform is a game changer for infrastructure management, the ability to maintain everything as code is a godsend and our entire organization is using it to manage thousands of resources across 10 teams.

During our 3 years of setting up this infrastructure, we've noticed one issue that has been hard to reconcile: The fact that Terraform is "all powerful" whilst it is the only acceptable way of changing our infra. These two concepts are constantly at odds with each other. But it doesn't have to be this way.

Terraform apply is really the dangerous operation. It affects the infra, it can delete crucial resources, it has "write" permissions. Terraform plan however, does not do anything like that. It just does a bunch of GET requests, compares to the state, and proposes changes, it has "read" permissions.

This mean Terraform plan should be runnable by any engineer at our company. They can make changes to the Terraform code on their git branch, run Terraform plan locally, evaluate their changes, iterate, and eventually create a pull request, that runs another Terraform plan, to be reviewed by the gatekeepers of our organization. Works great, everyone can develop with velocity, and the security is maintained.

However, its not that simple. There are a few issues that make Terraform plan nearly as privileged as Terraform apply. Resources contain secrets in their configurations. These secrets need to be read by Terraform plan from a datasource API and later during Terraform refresh read again from the resource API for change detection. This means the underlying accounts needs permissions to read secrets from these APIs. Additionally, these secrets are then kept in the Terraform state, which is also readable by Terraform plan.

During plan we'd want all operations to be as little privileged as possible, this way all engineers can develop and run `Terraform plan` without needing access to any secret values that could be misused for "write level" operations later.

### Attempted Solutions

# Our current workarounds

1. We let all applications that can, fetch directly from our vault. This meant Terraform does not need read permissions at all.
2. We set up "proxy" serverless functions, that are privileged to read from vault, and configure the secrets in the destination. Terraform apply has permissions to grant them access to the vault secrets, without having access itself. Terraform plan only has permissions to view the permissions of the serverless functions, not the secrets themselves.
3. We don't generate secrets with Terraform, or accept the fact they will be visible in the state. We experimented with serverless functions or null_resources to generate these. But this caused issues in the resource dependencies since the secrets aren't available yet at plan.

### Proposal

Please consider the following proposal for fixing this:

Read operations defined by the providers could be split between "plan" and "apply" time reads.

1. On plan time reads: Resources and datasources would have the option to omit the secrets and return metadata instead which uniquely identifies the secret value
2. On apply time reads: Resources and datasources would return the same data with the secrets included

This would require a change in the contract between Terraform and its providers. And 1 change to the way Terraform calls the providers during "read" operations.

Backwards compatibility should be easy enough, if the provider does not define separate plan and apply time read operations, just the plan time read operation would be called.

The information returned during the apply time read would have to be consistent with the earlier plan time read. However this should solvable possible in most situations. (eg. Metadata or a unique version identifier). Also there could be a check that all the other fields haven't changed in the meantime.

This also means the underlying platforms and APIs should support separate metadata and read permissions. This is already supported by eg Vault and GCP secrets.

# An example:

A datasource "vault_kv_secret_v2" is used to configure a resource "google_secret_manager_secret_version".

The 2 read operations will be referred to as "plan_read" and "apply_read"

In this example, both these resources have been updated to use the following endpoints:

- vault_kv_secret_v2 plan_read: https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#read-secret-metadata
- vault_kv_secret_v2 apply_read: https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#read-secret-version
- google_secret_manager_secret_version plan_read: https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets.versions/list or https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets/get
- google_secret_manager_secret_version apply_read: https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets.versions/get

Creation Plan (using low privileges):
1. Terraform calls plan_read of vault_kv_secret_v2, which will return only the metadata.
2. Terraform plans a create action and includes it in the `.tfplan`. This will include a reference to vault_kv_secret_v2 and the metadata (no secrets)

Creation Apply (using higher privileges):
1. Terraform calls apply_read apply_read of vault_kv_secret_v2, which will return the secret data
2. Terraform creates google_secret_manager_secret_version using the metadata
3. The Terraform state is updated with the new resource, which instead of the secret value, contains the metadata returned by the endpoint, as well as a reference to the datasource that is the source for the secret.
4. For the datasource the metadata is also stored.

Update/Read Plan:
1. Terraform calls plan_read of vault_kv_secret_v2, which will return only the metadata.
2. Terraform calls plan_read of google_secret_manager_secret_version, which will return only the metadata.
3. Terraform compares the metadata of both to the known metadata in the state
4. If the metadata is the same Terraform does nothing with it and continues.
5. If the metadata is different Terraform will include changes for all dependent resources in the state with a value that won't be known until apply.

Update: Apply
1. Terraform calls apply_read apply_read of vault_kv_secret_v2, which will return the secret data
2. Terraform updates google_secret_manager_secret_version using the metadata
3. The state is updated with the new metadata

# Our current workarounds

1. We let all applications that can, fetch directly from our vault. This meant Terraform does not need read permissions at all.
2. We set up "proxy" serverless functions, that are privileged to read from vault, and configure the secrets in the destination. Terraform apply has permissions to grant them access to the vault secrets, without having access itself. Terraform plan only has permissions to view the permissions of the serverless functions, not the secrets themselves.
3. We don't generate secrets with Terraform, or accept the fact they will be visible in the state. We experimented with serverless functions or null_resources to generate these. But this caused issues in the resource dependencies since the secrets aren't available yet at plan.

### References

- #516
- #32100

**Comment by rwblokzijl** [comment]
It seems Terraform has come closer to this in the last year than I could have hoped. Ephemeral values and resources are great! They are already allowing for some elimination of secrets from the statefile. And they're allowing running of plans with lower levels of privilege.

With the introduction of [write only attributes](https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/write-only-arguments) this has improved even further allowing secrets to be configured in resources without needing to add them to the state or plan files.

I don't what the current roadmap or vision of Ephemeral values looks like, but I'd like to add my 2 cents based on the usecases I described in this thread one year ago.

Currently we have the following problems still:

# 1 Reading a secret in Terraform and writing it to an resource required high privileges during plan time

With write only attributes these secrets no longer end up in the state. However, the Terraform plan pipelines still need permissions to read secrets en e.g. HashiCorp vault or google secret manager. This often means Terraform plans become highly privileged operations. Something that is not compatable with the GitOps pipelines that many companies have built around their Terraform stacks.

One way to solve this would be to allow "Ephemeral" resources to provide the data **only at apply** time with some flag.

However, this does break plannability. If Terraform cannot read the datasource at plan time it does not know whether to include the downstream resources as changed in the plan.

This could be solved by allowing Ephemeral resources to persist some "metadata" in the state. Secret manager like like HashiCorp vault or google secrets could use something like a "secret version" or a "last-updated" timestamp. Providers would implement a getMetadata function in the provider to get this. Terraform calls this at plan time and compares it to the state to know whether it should include the dependent resources in the plan.

# 2 Resources that generate secrets

There are many resources that output some kind of secret. For example [`random_password`](https://registry.terraform.io/providers/hashicorp/random/latest/docs/resources/password) and [`tls_private_key`](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key). These secrets are often generated in Terraform and passed to various services to authenticate with eachother. However these secrets also end up in statefiles and sometimes plan files. 

To solve this something like a "ReadOnceOnCreate" argument could be added to the attributes in the provider schema. When this is set it would make the attribute Ephemeral and only make it available during the apply when it is created.

This would allow users to store these secrets directly into a secret manager with a write-only argument without storing the value in the state.

Some of these resources exist only in the state (like random_password and tls_private_key) they would then only serve as "lifecycle management" of the secret allowing for easy rotations and updates downstream.

Another example is HashiCorp vault. When initializing, it generates a "root key" that allows the users to do the initial configuration. This init API endpoint can be called only once and there is NO option for subsequent reads of this secret. A resource for vault init currently does not exist, but one could be created with this mechanism.

#

I hope some things are coming down the pipeline to solve these scenarios. Based on the features added the past year I have the fullest confidence in the direction of this project :)

**Comment by crw** [comment]
@rwblokzijl Thanks for this update! It is appreciated.

**Comment by crw** [comment]
@rwblokzijl:

For 1, please check out https://developer.hashicorp.com/terraform/language/functions/terraform-applying -- this may offer a solution to your use case. If not, let us know. 

For 2, I believe this simply requires the providers to support the various functions required. For example see https://registry.terraform.io/providers/hashicorp/random/latest/docs/ephemeral-resources/password -- this exists today, but the tls equivalent does not yet. If that what you are looking for, you would want to open and follow issues on the repositories for each provider.

**Comment by kaitobq** [comment]
I think our use case is a concrete example of the remaining gap described here: ephemeral + write-only keeps the secret out of state/plan artifacts, but the terraform plan runner still needs permission to read/decrypt the secret source.

We tested this with:

- Terraform v1.15.3
- hashicorp/aws v6.47.0
- carlpett/sops v1.4.1
- ephemeral "sops_file" as the secret source
- aws_secretsmanager_secret_version.secret_string_wo as the only destination for the decrypted value
- secret_string_wo_version unchanged

Configuration shape:
```
ephemeral "sops_file" "github_app" {
  source_file = "sops/pr_plan_github_app.sops.yaml"
}

resource "aws_secretsmanager_secret_version" "github_app" {
  secret_id                = aws_secretsmanager_secret.github_app.id
  secret_string_wo         = ephemeral.sops_file.github_app.data["private_key"]
  secret_string_wo_version = 3
}
```
We removed kms:Decrypt from the role used only for PR terraform plan, expecting that Terraform might not need to open the ephemeral value because it only flows into a write-only argument and the write-only version did not change.

The plan still failed before producing a diff:
```
Error reading data

  with ephemeral.sops_file.pr_plan_github_app,
  on codebuild_pr_plan.tf line 13, in ephemeral "sops_file" "pr_plan_github_app":
  13: ephemeral "sops_file" "pr_plan_github_app" {

Error decrypting sops file: Error getting data key: 0 successful groups required, got 0
```
A second ephemeral "sops_file" used the same pattern and failed the same way. So the plan role still needs KMS decrypt permission for the encrypted SOPS file, even though the decrypted value is only passed to a write-only argument and is not stored in state or plan output.

terraform.applying does not help here. As long as an ephemeral block is declared and referenced anywhere in an expression, Terraform opens it during plan — which is where the decrypt happens — regardless of how the resulting value is gated. We verified two variants against the same de-privileged plan role (no kms:Decrypt):

1. Gate the value with terraform.applying, keeping the ephemeral declared (a non-null placeholder is used on the plan side so the write-only argument and its version are both specified):
secret_string_wo = terraform.applying ? ephemeral.sops_file.github_app.data["private_key"] : "placeholder"
1. → Error reading data ... ephemeral.sops_file ... Error decrypting sops file. The ephemeral is opened at plan even though, with terraform.applying = false, its value is never selected. The plan still requires kms:Decrypt.
2. Try to make the ephemeral apply-only by gating its count:
```
ephemeral "sops_file" "github_app" {
  count       = terraform.applying ? 1 : 0
  source_file = "..."
}
```
2. → Invalid count argument: "The given count is derived from an ephemeral value, which means that Terraform cannot persist it between plan/apply rounds. Use only non-ephemeral values to specify the number of resource instances."

The only way we found to drop kms:Decrypt from the plan role is to remove the ephemeral declaration from the plan graph entirely, using a non-ephemeral variable to drive count and passing a different value per pipeline (plan stage passes 0, apply stage passes 1):
```
ephemeral "sops_file" "github_app" {
  count       = var.include_secret_material ? 1 : 0   # plan: false → 0, apply: true → 1
  source_file = "..."
}

resource "aws_secretsmanager_secret_version" "github_app" {
  secret_string_wo         = var.include_secret_material ? ephemeral.sops_file.github_app[0].data["private_key"] : "placeholder"
  secret_string_wo_version = 3
}
```
This runs the plan with ReadOnlyAccess only (no kms:Decrypt) and produces no spurious diff, because the value flows only into a write-only argument and secret_string_wo_version is unchanged. The tradeoff is that the plan deliberately drops the ephemeral, so it cannot use any plan-time information from the secret source — it relies entirely on the manually-maintained secret_string_wo_version.

What would actually close the gap is closer to the proposal in this issue: a way for Terraform/providers to use only non-secret metadata during plan, and defer opening/reading the actual ephemeral secret until apply, while still preserving enough information to decide whether dependent write-only resources need to change. In our case, metadata such as secret_string_wo_version plus SOPS file identity/version would be enough for the PR plan; the decrypted SOPS value should only be needed in the trusted apply context.

Should this be tracked here, or would you prefer a more specific issue for apply-only ephemeral reads into write-only arguments?

**Comment by jbardin** [comment]
Hi @kaitobq,

Yes, unfortunately this issue actually contains two very different feature requests. Both of these are already known problem areas that need to be looked into though, so I think we can just leave the issue as-is for now as a placeholder and still capture the existing 👍's .
